src/envoy/http/alpn/alpn_test.cc

```diff
@@ -31,55 +34,139 @@ namespace {
+
+    auto config =
+        std::make_shared<AlpnFilterConfig>(proto_config, cluster_manager_);
     auto filter = std::make_unique<AlpnFilter>(config);
     filter->setDecoderFilterCallbacks(callbacks_);
     return filter;
   }
 
  protected:
+  FilterConfig::Protocol getProtocol(Http::Protocol protocol) {
+    switch (protocol) {
+      case Http::Protocol::Http10:
+        return FilterConfig::Protocol::FilterConfig_Protocol_HTTP10;
+      case Http::Protocol::Http11:
+        return FilterConfig::Protocol::FilterConfig_Protocol_HTTP11;
+      case Http::Protocol::Http2:
+        return FilterConfig::Protocol::FilterConfig_Protocol_HTTP2;
+      default:
+        NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+    }
+  }
+
   NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  NiceMock<Upstream::MockClusterManager> cluster_manager_;
+  std::shared_ptr<Upstream::MockThreadLocalCluster> fake_cluster_{
+      std::make_shared<NiceMock<Upstream::MockThreadLocalCluster>>()};
+  std::shared_ptr<Upstream::MockClusterInfo> cluster_info_{
+      std::make_shared<NiceMock<Upstream::MockClusterInfo>>()};
   Http::TestHeaderMapImpl headers_;
 };
 
-TEST_F(AlpnFilterTest, NoAlpnOverride) {
+TEST_F(AlpnFilterTest, OverrideAlpnUseDownstreamProtocol) {
   NiceMock<StreamInfo::MockStreamInfo> stream_info;
   ON_CALL(callbacks_, streamInfo()).WillByDefault(ReturnRef(stream_info));
-  auto filter = makeDefaultFilter();
-  EXPECT_CALL(stream_info, filterState()).Times(0);
-  EXPECT_EQ(filter->decodeHeaders(headers_, false),
-            Http::FilterHeadersStatus::Continue);
+  AlpnOverride alpn = {{Http::Protocol::Http10, {"foo", "bar"}},
+                       {Http::Protocol::Http11, {"baz"}},
+                       {Http::Protocol::Http2, {"qux"}}};
+  auto filter = makeAlpnOverrideFilter(alpn);
+
+  ON_CALL(cluster_manager_, get(_)).WillByDefault(Return(fake_cluster_.get()));
+  ON_CALL(*fake_cluster_, info()).WillByDefault(Return(cluster_info_));
+  ON_CALL(*cluster_info_, upstreamHttpProtocol(_))
+      .WillByDefault([](absl::optional<Http::Protocol> protocol) {
+        return protocol.value();
+      });
+
+  auto protocols = {Http::Protocol::Http10, Http::Protocol::Http11,
+                    Http::Protocol::Http2};
+  for (const auto p : protocols) {
+    EXPECT_CALL(stream_info, protocol()).WillOnce(Return(p));
+    Envoy::StreamInfo::FilterStateImpl filter_state;
+    EXPECT_CALL(stream_info, filterState()).WillOnce(ReturnRef(filter_state));
+    EXPECT_EQ(filter->decodeHeaders(headers_, false),
+              Http::FilterHeadersStatus::Continue);
+    EXPECT_TRUE(filter_state.hasData<Network::ApplicationProtocols>(
+        Network::ApplicationProtocols::key()));
+    auto alpn_override = filter_state
+                             .getDataReadOnly<Network::ApplicationProtocols>(
+                                 Network::ApplicationProtocols::key())
+                             .value();
+
+    EXPECT_EQ(alpn_override, alpn[p]);
```

Comment:
let's split it into two expectation:
 hasValue() and value EQ

Reply:
I checked `hasData` above `getDataReadOnly`